### PR TITLE
Add ingress rule support to the GCE provider

### PR DESCRIPTION
**Pull request by wallyworld**
The GCE provider is updating to support the new ingress rules with source CIDRs.
The change looks bigger than is it because git decided the file move (ruleset.go from network -> provider/gce/google) was an add/remove.

The existing way the GCE provider managed port ranges was to explode them into individual ports. This is allowed for in the new implementation - we read in existing rules and collapse the port ranges. When we write out any changes, or create new rules port ranges are used.

We now create different firewalls corresponding to distinct CIDR sets to fit in with the GCE model. For the default 0.0.0.0/0 case, there is no change. As new CIDRs are added and removed, the firewall name will change to allow for distinct port ranges with their own CIDR set.

There's also a drive by tweak to the azure provider

QA
bootstrap GCE with older juju
expose an application
open ports 3100-3110
observe that the firewall rules are individual ports
upgrade juju
open ports 3120-3130
observe that the firewall is updated and now has port ranges 3100-3110,3120-3130
close ports 3100-3110
observe that the firewall is updated and now has port ranges 3120-3130

### CODE REVIEW

**Comment by jameinel** [comment]
Given that both open and close essentially generate a new list of firewall rules and then figure out what needs adding and deleting, would it be possible to share more of the implementation. Eg common helper to slurp up existing rules, each modifies the set separately, common helper to turn the set into new rules, compare it with existing rules and write them out again?

Given what I just read about GCE firewall rules I'm a little concerned that when you call OpenPorts for MySQL that it will also open those ports for all machines that are in the same "network". I didn't quite see how we chose which firewall rules apply to which machine and the rules we seem to be creating are all Source based.

I'll try to look deeper.

**Comment by wallyworld** [comment]
I have intentions of generalising some for the logic for dealing with firewall rules.... but, I would rather land this bespoke and wait until we are in a position to modify other providers. It's easier to do well with more than one use case. Also, the other providers use a slightly different model and semantics so we still need to see what's applicable. My desire is to unblock the ability to start looking at the firewall worker as a main priority.

The implementation does only filter based on the source of the packets, that is true. And right now as a first step, the source will be set to packets originating from any network in the consuming model.  As the Juju modelling improves over time. and some of the network capabilities like binding/offering endpoints to spaces come along, we can tighten it up. Or am I missing the point of your concern?

File: provider/gce/google/conn_network.go [hunk 1]
```diff
@@ -81,33 +199,71 @@ func (gce Connection) OpenPorts(fwname string, rules ...network.IngressRule) err
...
+
+		existingFirewallName, sourceCIDRMatch := matchSourceCIDRs(currentFirewallRules, currentRuleSet, inputCidrs)
+		if !sourceCIDRMatch {
+			// We already know ports don't match, so if CIDRs don't match either, we either
+			// have a partial match or no match.
+			// No matches are a no-np. Partial matches might required splitting firewall rules
+			// which is not supported at the moment. We'll return an error as it's better to
+			// be overly cautious than accidentally leave ports open. The issue should occur
+			// in practice unless people have manually played with the firewall rules.
+			return errors.NotSupportedf("closing port(s) %v+ over non-matching rules", rules)
```
Comment (axw) [inline provider/gce/google/conn_network.go]: is that supposed to be %+v?

File: provider/gce/google/conn_network.go [hunk 2]
```diff
@@ -81,33 +199,71 @@ func (gce Connection) OpenPorts(fwname string, rules ...network.IngressRule) err
...
+			continue
+		}
+
+		existingFirewallName, sourceCIDRMatch := matchSourceCIDRs(currentFirewallRules, currentRuleSet, inputCidrs)
+		if !sourceCIDRMatch {
+			// We already know ports don't match, so if CIDRs don't match either, we either
+			// have a partial match or no match.
+			// No matches are a no-np. Partial matches might required splitting firewall rules
+			// which is not supported at the moment. We'll return an error as it's better to
+			// be overly cautious than accidentally leave ports open. The issue should occur
```
Comment (axw) [inline provider/gce/google/conn_network.go]: s/should/shouldn't/

File: provider/gce/google/conn_network.go [hunk 3]
```diff
@@ -81,33 +199,71 @@ func (gce Connection) OpenPorts(fwname string, rules ...network.IngressRule) err
...
+				return errors.Annotatef(err, "closing port(s) %+v", rules)
+			}
+			continue
+		}
+
+		existingFirewallName, sourceCIDRMatch := matchSourceCIDRs(currentFirewallRules, currentRuleSet, inputCidrs)
+		if !sourceCIDRMatch {
+			// We already know ports don't match, so if CIDRs don't match either, we either
+			// have a partial match or no match.
+			// No matches are a no-np. Partial matches might required splitting firewall rules
```
Comment (axw) [inline provider/gce/google/conn_network.go]: s/np-np/no-op/

Comment (axw) [inline provider/gce/google/conn_network.go]: s/required/require/

File: provider/gce/google/raw.go [hunk 4]
```diff
@@ -115,18 +115,23 @@ func (rc *rawConn) RemoveInstance(projectID, zone, id string) error {
...
 	}
 
 	if len(firewallList.Items) == 0 {
-		return nil, errors.NotFoundf("firewall %q", name)
+		return nil, errors.NotFoundf("firewall %q", namePrefix)
 	}
-	return firewallList.Items[0], nil
+	var result []*compute.Firewall
+	for _, fw := range firewallList.Items {
+		if fw.Name == namePrefix || strings.HasPrefix(fw.Name, namePrefix) {
```
Comment (axw) [inline provider/gce/google/raw.go]: == is redundant

File: provider/gce/google/conn_network.go [hunk 5]
```diff
@@ -22,57 +25,172 @@ func (gce Connection) IngressRules(fwname string) ([]network.IngressRule, error)
...
+			sourceRanges = []string{"0.0.0.0/0"}
+		}
+		portsForProtocol := make(map[string]set.Ints)
+		for _, allowed := range firewall.Allowed {
+			for _, portRangeStr := range allowed.Ports {
+				portRange, err := network.ParsePortRange(portRangeStr)
+				if err != nil {
+					return rules, errors.Annotate(err, "bad ports from GCE")
+				}
+				for p := portRange.FromPort; p <= portRange.ToPort; p++ {
```
Comment (wallyworld) [inline provider/gce/google/conn_network.go]: Yeah, I reused what was there previously, but have now reworked it.

File: provider/gce/google/conn_network.go [hunk 6]
```diff
@@ -22,57 +25,172 @@ func (gce Connection) IngressRules(fwname string) ([]network.IngressRule, error)
...
+}
 
-	// Update an existing firewall.
-	newPortsSet := currentPortsSet.Union(inputPortsSet)
-	firewall := firewallSpec(fwname, newPortsSet)
-	if err := gce.raw.UpdateFirewall(gce.projectID, fwname, firewall); err != nil {
-		return errors.Annotatef(err, "opening port(s) %+v", rules)
+// matchIgnoreSourceCIDRs returns the matching firewall name and true if any of the current
+// firewall ports (ignoring CIDR values) matches the ports to match.
+func matchIgnoreSourceCIDRs(currentPorts map[string]protocolPorts, match protocolPorts) (string, bool) {
```
Comment (wallyworld) [inline provider/gce/google/conn_network.go]: done

File: provider/gce/google/conn_network.go (hunk 3 above)
Comment (wallyworld) [inline provider/gce/google/conn_network.go]: done

File: provider/gce/google/conn_network.go (hunk 2 above)
Comment (wallyworld) [inline provider/gce/google/conn_network.go]: done

File: provider/gce/google/conn_network.go (hunk 1 above)
Comment (wallyworld) [inline provider/gce/google/conn_network.go]: fixed

File: network/portrange.go [hunk 7]
```diff
@@ -196,3 +196,28 @@ func parsePortRange(portRange string) (PortRange, error) {
 	}
 	return result, nil
 }
+
+// CombinePortRanges groups together all port ranges according to
```
Comment (axw) [inline network/portrange.go]: please call out in flashing capital letters that this only works on non-overlapping sets

Comment (wallyworld) [inline network/portrange.go]: done

File: network/portrange.go [hunk 8]
```diff
@@ -196,3 +196,30 @@ func parsePortRange(portRange string) (PortRange, error) {
...
+// protocol, and then combines then into contiguous port ranges.
+// NOTE: Juju only allows it's model to contain non-overlapping port ranges.
```
Comment (axw) [inline network/portrange.go]: s/it's/its/

Comment (wallyworld) [inline network/portrange.go]: I hate myself for that mistake. I'm normally anal about that stuff.

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**